File: api/Enkl.Api/Services/ProjectSettingsSerializer.cs
using System.Text.Json;
using Enkl.Api.Dtos;

namespace Enkl.Api.Services;

/// <summary>
/// Shared camelCase (de)serialization for Project.HeaderButtonVisibilityJson. Defaults mirror
/// normalizeHeaderButtonVisibility (src/js/storage.js) exactly: every field is opt-out (defaults to
/// true, so a missing/corrupted value never silently hides something the user never chose to hide)
/// except Workflow and ChangeAuditing, which are opt-in (default false, so a missing/corrupted value
/// never silently starts enforcing/recording something the user never asked for).
/// </summary>
public static class ProjectSettingsSerializer
{
    private static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web);

    public static string Serialize(ProjectSettingsDto settings) => JsonSerializer.Serialize(settings, Options);

    public static ProjectSettingsDto Parse(string? json)
    {
        JsonDocument? doc = null;
        try
        {
            if (!string.IsNullOrWhiteSpace(json)) doc = JsonDocument.Parse(json);
        }
        catch (JsonException)
        {
            // Corrupted/garbled JSON falls through to defaults below, same as the client-side guard.
        }

        bool Get(string name, bool defaultValue)
        {
            if (doc is not null && doc.RootElement.TryGetProperty(name, out var val) &&
                (val.ValueKind == JsonValueKind.True || val.ValueKind == JsonValueKind.False))
            {
                return val.ValueKind == JsonValueKind.True;
            }
            return defaultValue;
        }

        var result = new ProjectSettingsDto(
            Documents: Get("documents", true),
            Risks: Get("risks", true),
            Decisions: Get("decisions", true),
            Health: Get("health", true),
            Principles: Get("principles", true),
            Objectives: Get("objectives", true),
            TeamsCommittees: Get("teamsCommittees", true),
            Workflow: Get("workflow", false),
            TimeTracking: Get("timeTracking", true),
            ChangeAuditing: Get("changeAuditing", false),
            SubTasks: Get("subTasks", true));

        doc?.Dispose();
        return result;
    }
}

File: mariadb-api/src/Services/ProjectSettingsSerializer.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

/**
 * Ported from Services/ProjectSettingsSerializer.cs. Defaults mirror normalizeHeaderButtonVisibility
 * (src/js/storage.js) exactly: every field is opt-out (defaults true) except workflow,
 * changeAuditing and retrospective, which are opt-in (default false) — a missing/corrupted value
 * must never silently start enforcing/recording/showing something the user never asked for. Keys
 * are camelCase to match both the frontend's own field names and the "changeAuditing" key
 * TaskService::isChangeAuditingEnabled reads from this same column.
 *
 * @phpstan-type ProjectSettings array{documents:bool,risks:bool,decisions:bool,health:bool,principles:bool,objectives:bool,teamsCommittees:bool,workflow:bool,timeTracking:bool,changeAuditing:bool,subTasks:bool,retrospective:bool,strategy:bool}
 */
final class ProjectSettingsSerializer
{
    private const DEFAULTS = [
        'documents' => true,
        'risks' => true,
        'decisions' => true,
        'health' => true,
        'principles' => true,
        'objectives' => true,
        'teamsCommittees' => true,
        'workflow' => false,
        'timeTracking' => true,
        'changeAuditing' => false,
        'subTasks' => true,
        // Opt-in, like workflow: brand-new functionality nobody has configured yet, so a
        // missing/corrupted value must never silently turn it on.
        'retrospective' => false,
        // Opt-in, like workflow/retrospective: a missing/corrupted value must never silently turn on
        // a module the project never asked for.
        'strategy' => false,
    ];

    public static function serialize(array $settings): string
    {
        $result = [];
        foreach (self::DEFAULTS as $key => $default) {
            $result[$key] = (bool) ($settings[$key] ?? $default);
        }
        return json_encode($result);
    }

    /** @return ProjectSettings */
    public static function parse(?string $json): array
    {
        $decoded = [];
        if ($json !== null && $json !== '') {
            $tmp = json_decode($json, true);
            if (is_array($tmp)) {
                $decoded = $tmp;
            }
        }

        $result = [];
        foreach (self::DEFAULTS as $key => $default) {
            $value = $decoded[$key] ?? null;
            $result[$key] = is_bool($value) ? $value : $default;
        }
        return $result;
    }
}
